### PR TITLE
satellite/gracefulexit: add piece hash validation for successful transfer

### PR DESCRIPTION
**Pull request by navillasa**
What: 
- verify that the satellite signed the original order limit
- verify that the original piece hash and replacement piece hash match
- verify that the public key from the original order limit signed the original piece hash
- verify that the replacement node signed the replacement piece hash

Why:
- need piece hash validation for successful transfer
https://storjlabs.atlassian.net/browse/V3-2874

Please describe the tests:
need tests still!!!!
 - Test 1:
 - Test 2:
 
Please describe the performance impact:

## Code Review Checklist (to be filled out by reviewer)
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

### CODE REVIEW

File: satellite/gracefulexit/endpoint.go [hunk 1]
```diff
@@ -398,43 +412,103 @@ func (endpoint *Endpoint) processIncomplete(ctx context.Context, stream processS
...
+
+	// get peerID and signee for new storage node
+	peerID, err := endpoint.peerIdentities.Get(ctx, receivingNodeID)
 	if err != nil {
 		return Error.Wrap(err)
 	}
+	signee := signing.SigneeFromPeerIdentity(peerID)
+
+	// verify that the new node signed the replacement piece hash
+	err = signing.VerifyPieceHashSignature(ctx, signee, replacementPieceHash)
```
Comment (egonelbre) [inline satellite/gracefulexit/endpoint.go]: There seems to be missing `piece id` verification between replacement piece hash and replacement order limit. Would it be essential?

File: satellite/gracefulexit/endpoint_test.go [hunk 2]
```diff
@@ -97,47 +128,287 @@ func TestSuccess(t *testing.T) {
...
-			if errs.Is(err, io.EOF) {
-				// Done
-				break
-			}
-			require.NoError(t, err)
-
-			switch m := response.GetMessage().(type) {
-			case *pb.SatelliteMessage_TransferPiece:
-				require.NotNil(t, m)
+	for _, tt := range []struct {
```
Comment (egonelbre) [inline satellite/gracefulexit/endpoint_test.go]: Rewrite this without table tests.

File: satellite/gracefulexit/endpoint_test.go [hunk 3]
```diff
@@ -161,9 +437,14 @@ func testTransfers(t *testing.T, objects int, verifier func(ctx *testcontext.Con
 			require.NoError(t, err)
 		}
 		// check that there are no exiting nodes.
-		exitingNodeIDs, err := satellite.DB.OverlayCache().GetExitingNodes(ctx)
-		require.NoError(t, err)
-		require.Len(t, exitingNodeIDs, 0)
+		for {
```
Comment (egonelbre) [inline satellite/gracefulexit/endpoint_test.go]: Why is this a for loop?

Comment (egonelbre) [reply satellite/gracefulexit/endpoint_test.go]: Maybe it's missing something like `satellite.XYZ.Cycle.Wait()` or `satellite.XYZ.Cycle.TriggerWait()`

File: satellite/gracefulexit/endpoint.go [hunk 4]
```diff
@@ -398,43 +412,103 @@ func (endpoint *Endpoint) processIncomplete(ctx context.Context, stream processS
...
+	err = signing.VerifyUplinkPieceHashSignature(ctx, originalOrderLimit.UplinkPublicKey, originalPieceHash)
+	if err != nil {
+		failedCount++
+		transferQueueItem.FailedCount = &failedCount
+		updateErr := endpoint.db.UpdateTransferQueueItem(ctx, *transferQueueItem)
+		return Error.Wrap(errs.Combine(updateErr, err))
+	}
+
+	// TODO add nil checks/figure out a better way to get the receiving node ID
+	receivingNodeID := transfer.satelliteMessage.GetTransferPiece().GetAddressedOrderLimit().GetLimit().StorageNodeId
```
[the remark this answers is not in the document]
Comment (navillasa) [reply satellite/gracefulexit/endpoint.go]: looks like the original order limit only involves the exiting node, and doesn't have any information about the new receiving node.

File: satellite/gracefulexit/endpoint.go [hunk 5]
```diff
@@ -398,43 +412,103 @@ func (endpoint *Endpoint) processIncomplete(ctx context.Context, stream processS
...
+		return Error.Wrap(err)
 	}
 
-	transferQueueItem, err := endpoint.db.GetTransferQueueItem(ctx, nodeID, transfer.path)
+	failedCount := 0
+
+	originalOrderLimit := message.Succeeded.GetOriginalOrderLimit()
+	if originalOrderLimit == nil {
+		failedCount++
+		transferQueueItem.FailedCount = &failedCount
```
[the remark this answers is not in the document]
Comment (navillasa) [reply satellite/gracefulexit/endpoint.go]: I think this should probably be addressed in another PR but agree it's clunky.

File: satellite/gracefulexit/endpoint.go [hunk 6]
```diff
@@ -398,43 +412,103 @@ func (endpoint *Endpoint) processIncomplete(ctx context.Context, stream processS
...
+		updateErr := endpoint.db.UpdateTransferQueueItem(ctx, *transferQueueItem)
+		return Error.Wrap(errs.Combine(updateErr, Error.New("Original order limit cannot be nil.")))
+	}
+	originalPieceHash := message.Succeeded.GetOriginalPieceHash()
+	if originalPieceHash == nil {
+		failedCount++
+		transferQueueItem.FailedCount = &failedCount
+		updateErr := endpoint.db.UpdateTransferQueueItem(ctx, *transferQueueItem)
```
[the remark this answers is not in the document]
Comment (navillasa) [reply satellite/gracefulexit/endpoint.go]: I'm taking the `FailedCount` updates out of this PR for now

File: satellite/gracefulexit/endpoint.go [hunk 7]
```diff
@@ -169,6 +179,10 @@ func (endpoint *Endpoint) doProcess(stream processStream) (err error) {
 		if err != nil {
 			return Error.Wrap(err)
 		}
+		err = endpoint.db.IncrementProgress(ctx, nodeID, 0, 0, 0)
+		if err != nil {
+			return Error.Wrap(err)
```
[the remark this answers is not in the document]
Comment (navillasa) [reply satellite/gracefulexit/endpoint.go]: added these

File: satellite/gracefulexit/endpoint.go [hunk 8]
```diff
@@ -398,43 +412,103 @@ func (endpoint *Endpoint) processIncomplete(ctx context.Context, stream processS
...
 
 	pieceID := message.Succeeded.OriginalPieceId
-	endpoint.log.Debug("transfer succeeded", zap.Stringer("piece ID", pieceID))
-
-	// TODO validation
 
 	transfer, ok := pending.get(pieceID)
 	if !ok {
-		endpoint.log.Debug("could not find transfer message in pending queue. skipping.", zap.Stringer("piece ID", pieceID))
+		endpoint.log.Error("could not find transfer message in pending queue", zap.Stringer("piece ID", pieceID))
```
[the remark this answers is not in the document]
Comment (navillasa) [reply satellite/gracefulexit/endpoint.go]: not sure about this - changed it to an error log since I saw the TODO requested an error returned. I guess the log could be taken out now, but I'm not sure how important it is. maybe @ethanadams has an opinion?

File: satellite/gracefulexit/endpoint.go [hunk 9]
```diff
@@ -169,6 +179,10 @@ func (endpoint *Endpoint) doProcess(stream processStream) (err error) {
 		if err != nil {
```
[the remark this answers is not in the document]
Comment (mobyvb) [reply satellite/gracefulexit/endpoint.go]: That could be an issue. I made a PR to address this concern: https://github.com/storj/storj/pull/3357

File: satellite/gracefulexit/endpoint_test.go [hunk 10]
```diff
@@ -149,6 +378,11 @@ func testTransfers(t *testing.T, objects int, verifier func(ctx *testcontext.Con
 
 		satellite.GracefulExit.Chore.Loop.Pause()
 
+		storageNodes := make(map[storj.NodeID]*storagenode.Peer)
```
Comment (mobyvb) [inline satellite/gracefulexit/endpoint_test.go]: nit - this can be just the identities. It doesn't need to be the whole peer

File: satellite/gracefulexit/endpoint.go [hunk 11]
```diff
@@ -398,43 +412,103 @@ func (endpoint *Endpoint) processIncomplete(ctx context.Context, stream processS
...
 	transfer, ok := pending.get(pieceID)
 	if !ok {
-		endpoint.log.Debug("could not find transfer message in pending queue. skipping.", zap.Stringer("piece ID", pieceID))
+		endpoint.log.Error("could not find transfer message in pending queue", zap.Stringer("piece ID", pieceID))
+		return Error.New("could not find transfer message in pending queue")
+	}
 
-		// TODO we should probably error out here so we don't get stuck in a loop with a SN that is not behaving properly
+	// TODO add nil check for getting path
+	transferQueueItem, err := endpoint.db.GetTransferQueueItem(ctx, exitingNodeID, transfer.path)
```
[the remark this answers is not in the document]
Comment (mobyvb) [reply satellite/gracefulexit/endpoint.go]: @navillasa I would move this line back down to where it was before (right before the checks for FailedCount)

File: satellite/gracefulexit/endpoint.go (hunk 1 above)
Comment (mobyvb) [reply satellite/gracefulexit/endpoint.go]: It wouldn't hurt to have but it's not essential. We derive the piece ID for the replacement order limit the same way Nat did above for `calculatedNewPieceID`.

Comment (egonelbre) [reply satellite/gracefulexit/endpoint.go]: Yes, but the exiting storage-node could provide a previously submitted piece hash.

Comment (mobyvb) [reply satellite/gracefulexit/endpoint.go]: We get `receivingNodeID` from the replacement order limit, and I think if the receiving node has signed the hash, and the piece ID on the hash matches what we calculate it should be, everything should be okay.
But I could be overlooking something and it wouldn't hurt to also check that the piece ID of the order limit is the same as the piece ID of the hash.

File: satellite/gracefulexit/endpoint.go [hunk 12]
```diff
@@ -393,53 +418,101 @@ func (endpoint *Endpoint) processIncomplete(ctx context.Context, stream processS
...
 	defer mon.Task()(&ctx)(&err)
-	if message.Succeeded.GetOriginalOrderLimit() == nil {
-		return Error.New("original order limit cannot be nil.")
+
+	originalPieceID := message.Succeeded.OriginalPieceId
+
+	transfer, ok := pending.get(originalPieceID)
+	if !ok {
+		endpoint.log.Error("could not find transfer message in pending queue", zap.Stringer("piece ID", originalPieceID))
+		return Error.New("could not find transfer message in pending queue")
```
Comment (VinozzZ) [inline satellite/gracefulexit/endpoint.go]: nit - `tranfer message` -> `transfer item` ?

File: satellite/gracefulexit/endpoint.go (hunk 1 above)
Comment (egonelbre) [reply satellite/gracefulexit/endpoint.go]: If I send you a properly signed replacement piece hash for a wrong piece id, what would happen currently?

Comment (mobyvb) [reply satellite/gracefulexit/endpoint.go]: This check above should fail
```
calculatedNewPieceID := transfer.rootPieceID.Derive(receivingNodeID, transfer.pieceNum)
if calculatedNewPieceID != replacementPieceHash.PieceId {
	return ErrInvalidReplacementPieceID
}
```

File: satellite/gracefulexit/endpoint.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (mobyvb) [reply satellite/gracefulexit/endpoint.go]: I think in the worst case scenario, if a node connects to `doProcess` in parallel, the satellite would likely send multiple piece transfer requests for the same pieces to the node. Which means that we might end up with duplicate pieces/order limits across multiple "recipient" nodes.
But on the satellite side, we should be able to handle this so that only the first transfer is actually successful, and the others are ignored. And the duplicate pieces would eventually be picked up by garbage collection.
In other words, it is not ideal, but I don't think it is necessarily a breaking issue. What do you think?

File: satellite/gracefulexit/endpoint.go [hunk 13]
```diff
@@ -432,33 +464,81 @@ func (endpoint *Endpoint) processIncomplete(ctx context.Context, stream processS
...
 	defer mon.Task()(&ctx)(&err)
-	if message.Succeeded.GetOriginalOrderLimit() == nil {
-		return Error.New("original order limit cannot be nil.")
+
+	originalPieceID := message.Succeeded.OriginalPieceId
+
+	transfer, ok := pending.get(originalPieceID)
+	if !ok {
+		endpoint.log.Error("could not find transfer item in pending queue", zap.Stringer("piece ID", originalPieceID))
+		return Error.New("could not find transfer item in pending queue")
```
Comment (VinozzZ) [inline satellite/gracefulexit/endpoint.go]: Should all the errors returned by `handleSucceeded` be wrapped with rpc error?

Comment (VinozzZ) [reply satellite/gracefulexit/endpoint.go]: Maybe instead of directly returning rpc error from this method, we can create error class and return that. Then on line 335, we can have a switch statement to return corresponding rpc error based on different error classes
